lib/WooCommerce/Subscription.php

```diff
@@ -0,0 +1,112 @@
+<?php
+namespace MailPoet\WooCommerce;
+
+use MailPoet\Models\Segment;
+use MailPoet\Models\Subscriber;
+use MailPoet\Models\SubscriberSegment;
+use MailPoet\Settings\SettingsController;
+use MailPoet\Subscribers\Source;
+use MailPoet\Util\Helpers;
+use MailPoet\WP\Functions as WPFunctions;
+
+class Subscription {
+  const CHECKOUT_OPTIN_INPUT_NAME = 'mailpoet_woocommerce_checkout_optin';
+  const OPTIN_ENABLED_SETTING_NAME = 'woocommerce.optin_on_checkout.enabled';
+  const OPTIN_MESSAGE_SETTING_NAME = 'woocommerce.optin_on_checkout.message';
+
+  /** @var SettingsController */
+  private $settings;
+
+  /** @var WPFunctions */
+  private $wp;
+
+  function __construct(
+    SettingsController $settings,
+    WPFunctions $wp
+  ) {
+    $this->settings = $settings;
+    $this->wp = $wp;
+  }
+
+  function extendWooCommerceCheckoutForm() {
+    $input_name = self::CHECKOUT_OPTIN_INPUT_NAME;
+    $checked = $this->isCurrentUserSubscribed();
+    $label_string = $this->settings->get(self::OPTIN_MESSAGE_SETTING_NAME);
+    $template = $this->wp->applyFilters(
+      'mailpoet_woocommerce_checkout_optin_template',
+      $this->getSubscriptionField($input_name, $checked, $label_string),
+      $input_name,
+      $checked,
+      $label_string
+    );
+    echo $template;
+  }
+
+  private function getSubscriptionField($input_name, $checked, $label_string) {
+    return '<p class="form-row">
+      <label class="woocommerce-form__label woocommerce-form__label-for-checkbox checkbox">
+      <input type="checkbox" class="woocommerce-form__input woocommerce-form__input-checkbox input-checkbox" name="'.$this->wp->escAttr($input_name).'" id="'.$this->wp->escAttr($input_name).'" ' . ($checked ? 'checked' : '') . ' />
+        <span class="woocommerce-terms-and-conditions-checkbox-text">' . $this->wp->escHtml($label_string) . '</label>
+    </p>';
+  }
+
+  private function isCurrentUserSubscribed() {
+    $subscriber = Subscriber::getCurrentWPUser();
+    if ($subscriber instanceof Subscriber) {
```

Comment:
I would use an early termination here so that we get rid of one level of nesting:
```php
if (!$subscriber instanceof Subscriber) {
  return false;
}
$wc_segment = Segment::getWooCommerceSegment();
...
```